### PR TITLE
koordlet: cgroup file read and write is privatized and consolidated into the resource executor package.

### PR DESCRIPTION
**Pull request by Re-Grh**
### Ⅱ. Does this pull request fix one issue?
fix https://github.com/koordinator-sh/koordinator/issues/1017

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/resmanager/cgroup_reconcile_test.go [hunk 1]
```diff
@@ -1467,3 +1470,37 @@ func writeMemToCgroupFile(parentDir string, qos *slov1alpha1.ResourceQOS, helper
 	helper.WriteCgroupFileContents(parentDir, system.MemoryPriority, strconv.FormatInt(*qos.MemoryQOS.Priority, 10))
 	helper.WriteCgroupFileContents(parentDir, system.MemoryOomGroup, strconv.FormatInt(*qos.MemoryQOS.OomKillGroup, 10))
 }
+
+//This function is only intended for testing functions within the current file. For specific read/write functionalities, please refer to the executor package.
```
Comment (saintube) [inline pkg/koordlet/resmanager/cgroup_reconcile_test.go:1474]: please left a space at the beginning

File: pkg/koordlet/resmanager/cgroup_reconcile_test.go [hunk 2]
```diff
@@ -1467,3 +1470,37 @@ func writeMemToCgroupFile(parentDir string, qos *slov1alpha1.ResourceQOS, helper
...
+	if dataStr == "" {
+		return nil, fmt.Errorf("EmptyValueError")
+	}
+	if dataStr == "max" {
+		// compatible with cgroup valued "max"
+		data := int64(math.MaxInt64)
+		klog.V(6).Infof("read %s and got str value, considered as MaxInt64", r.Path(cgroupTaskDir))
+		return &data, nil
+	}
+	Intdata, err := strconv.ParseInt(strings.TrimSpace(dataStr), 10, 64)
```
Comment (saintube) [inline pkg/koordlet/resmanager/cgroup_reconcile_test.go:1501]: follow the camel case and use the lowercase for the first letter of a local variable.
FYI: https://google.github.io/styleguide/go/

File: pkg/koordlet/resmanager/cgroup_reconcile_test.go [hunk 3]
```diff
@@ -1467,3 +1470,37 @@ func writeMemToCgroupFile(parentDir string, qos *slov1alpha1.ResourceQOS, helper
...
+	}
+
+	filePath := r.Path(cgroupTaskDir)
+	klog.V(5).Infof("read %s", filePath)
+
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	dataStr := strings.Trim(string(data), "\n")
```
Comment (saintube) [inline pkg/koordlet/resmanager/cgroup_reconcile_test.go:1490]: perhaps `dataStr := strings.Trim(string(data)), "\n ")` is better especially while reading `max  `.

File: pkg/koordlet/resourceexecutor/cgroup.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	// CgroupMaxSymbolStr only appears in cgroups-v2 files, we consider the value as MaxInt64
+	CgroupMaxSymbolStr string = "max"
+	// CgroupMaxValueStr math.MaxInt64; writing `memory.high` with this do the same as set as "max"
+	CgroupMaxValueStr string = "9223372036854775807"
+)
+
+const EmptyValueError string = "EmptyValueError"
+
+// CgroupFileWriteIfDifferent writes the cgroup file if current value is different from the given value.
+// TODO: moved into resourceexecutor package and marked as private.
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/cgroup.go:41]: the `TODO` can be removed. same below

File: pkg/koordlet/resourceexecutor/psi.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	}, nil
+}
+
+func readPSI(pressureFilePath string) (sysutil.PSIStats, error) {
+	fileContents, err := os.ReadFile(pressureFilePath)
+	if err != nil {
+		return sysutil.PSIStats{}, err
+	}
+	//todo: delete these logs after panic handled
+	klog.V(4).Infof("read psi file contents: %s", string(fileContents))
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/psi.go:54]: need to print the path

File: pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go [hunk 6]
```diff
@@ -131,8 +132,12 @@ func (b *bvtPlugin) ruleUpdateCb(pods []*statesinformer.PodMeta) error {
...
 		corev1.PodQOSGuaranteed, corev1.PodQOSBurstable, corev1.PodQOSBestEffort} {
 		bvtValue := r.getKubeQOSDirBvtValue(kubeQOS)
-		kubeQOSCgroupPath := koordletutil.GetKubeQosRelativePath(kubeQOS)
-		if err := sysutil.CgroupFileWrite(kubeQOSCgroupPath, sysutil.CPUBVTWarpNs, strconv.FormatInt(bvtValue, 10)); err != nil {
+		kubeQOSCgroupPath := koordletutil.GetPodQoSRelativePath(kubeQOS)
+		bvtUpdater, err := resourceexecutor.NewCommonCgroupUpdater(sysutil.CPUBVTWarpNs.ResourceType(), kubeQOSCgroupPath, strconv.FormatInt(bvtValue, 10))
+		if err != nil {
+			klog.Infof("bvtupdater creat failed, dir %v, error %v", kubeQOSCgroupPath, err)
+		}
+		if err := bvtUpdater.Update(); err != nil {
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go:140]: If we expect to reduce the audit logs in the runtime hook plugins (otherwise, we can leave a TODO here and do it in another PR), we may need an Executor to do cacheable updates.

File: pkg/koordlet/resmanager/cgroup_reconcile_test.go [hunk 7]
```diff
@@ -1467,3 +1470,37 @@ func writeMemToCgroupFile(parentDir string, qos *slov1alpha1.ResourceQOS, helper
...
+	if dataStr == "" {
+		return nil, fmt.Errorf("EmptyValueError")
+	}
+	if dataStr == "max" {
+		// compatible with cgroup valued "max"
+		data := int64(math.MaxInt64)
+		klog.V(6).Infof("read %s and got str value, considered as MaxInt64", r.Path(cgroupTaskDir))
+		return &data, nil
+	}
+	intData, err := strconv.ParseInt(strings.TrimSpace(dataStr), 10, 64)
```
Comment (saintube) [inline pkg/koordlet/resmanager/cgroup_reconcile_test.go:1501]: no need to trim space again

File: pkg/koordlet/resourceexecutor/cgroup.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	filePath := r.Path(cgroupTaskDir)
+	klog.V(5).Infof("read %s", filePath)
+
+	data, err := os.ReadFile(filePath)
+	return strings.Trim(string(data), "\n"), err
+}
+
+func readCgroupAndParseInt64(parentDir string, r sysutil.Resource) (int64, error) {
+	s, err := cgroupFileRead(parentDir, r)
+	if err != nil && sysutil.IsCgroupDirErr(err) {
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/cgroup.go:126]: can we simply use `if err != nil {}`

File: pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go [hunk 9]
```diff
@@ -143,7 +149,11 @@ func (b *bvtPlugin) ruleUpdateCb(pods []*statesinformer.PodMeta) error {
 		podKubeQOS := podMeta.Pod.Status.QOSClass
 		podBvt := r.getPodBvtValue(podQOS, podKubeQOS)
 		podCgroupPath := koordletutil.GetPodCgroupDirWithKube(podMeta.CgroupDir)
-		if err := sysutil.CgroupFileWrite(podCgroupPath, sysutil.CPUBVTWarpNs, strconv.FormatInt(podBvt, 10)); err != nil {
+		bvtUpdater, err := resourceexecutor.NewCommonCgroupUpdater(sysutil.CPUBVTWarpNs.ResourceType(), podCgroupPath, strconv.FormatInt(podBvt, 10))
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go:152]: perhaps `NewCommonCgroupUpdater(sysutil.CPUBVTWarpNsName, ...)`

File: pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go [hunk 10]
```diff
@@ -143,7 +149,11 @@ func (b *bvtPlugin) ruleUpdateCb(pods []*statesinformer.PodMeta) error {
...
+		if err != nil {
+			klog.Infof("bvtupdater creat failed, dir %v, error %v", podCgroupPath, err)
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go:154]: typo `creat`

File: pkg/koordlet/resourceexecutor/cgroup.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+
+func readCgroupAndParseUint64(parentDir string, r sysutil.Resource) (uint64, error) {
+	s, err := cgroupFileRead(parentDir, r)
+	if err != nil {
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/cgroup.go:146]: should not wrap the error since it can be ResourceUnsupportedErr or ResourceCgroupDirErr

File: pkg/koordlet/util/system/psi_file.go [hunk 12]
```diff
@@ -43,22 +51,8 @@ type PSIStats struct {
...
-	klog.V(4).Infof("read psi file contents: %s", string(fileContents))
-	stats, err := parsePSIStats(bytes.NewReader(fileContents))
-	if err != nil {
-		return PSIStats{}, err
-	}
-	return stats, nil
-}
-
 // parsePSIStats parses the specified file for pressure stall information.
-func parsePSIStats(r io.Reader) (PSIStats, error) {
```
Comment (zwzhang0107) [inline pkg/koordlet/util/system/psi_file.go:61]: other files and struct can also be moved into `resourceexecutor`

File: pkg/koordlet/util/system/cgroup.go [hunk 13]
```diff
@@ -68,31 +65,6 @@ func (m *MemoryStatRaw) Usage() int64 {
...
-	}
-	if value == currentValue || value == CgroupMaxValueStr && currentValue == CgroupMaxSymbolStr {
-		// compatible with cgroup valued "max"
-		klog.V(6).Infof("read before write %s and got str value, considered as MaxInt64", r.Path(cgroupTaskDir))
-		return nil
-	}
-	return CgroupFileWrite(cgroupTaskDir, r, value)
-}
-
 func IsCgroupPathExist(parentDir string, r Resource) (bool, string) {
```
Comment (zwzhang0107) [inline pkg/koordlet/util/system/cgroup.go:68]: mv to `resourceexecutor`

File: pkg/koordlet/util/container.go [hunk 14]
```diff
@@ -193,16 +193,6 @@ func GetContainerCurMemLimitBytes(podParentDir string, c *corev1.ContainerStatus
 	return strconv.ParseInt(strings.TrimSpace(string(rawContent)), 10, 64)
 }
 
-// GetContainerCurTasks gets the TIDs of the given container's cgroup.
-// DEPRECATED: use resourceexecutor.CgroupReader instead.
```
Comment (zwzhang0107) [inline pkg/koordlet/util/container.go:197]: check other deprecated func

File: pkg/koordlet/runtimehooks/protocol/container_context.go [hunk 15]
```diff
@@ -191,8 +191,10 @@ func (c *ContainerContext) injectForOrigin() {
 			klog.V(5).Infof("set container %v/%v/%v cpuset %v on cgroup parent %v",
 				c.Request.PodMeta.Namespace, c.Request.PodMeta.Name, c.Request.ContainerMeta.Name,
 				*c.Response.Resources.CPUSet, c.Request.CgroupParent)
-			audit.V(2).Container(c.Request.ContainerMeta.ID).Reason("runtime-hooks").Message(
-				"set container cpuset to %v", *c.Response.Resources.CPUSet).Do()
+			if updated {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/protocol/container_context.go:194]: independent pr

File: pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go [hunk 16]
```diff
@@ -131,10 +132,15 @@ func (b *bvtPlugin) ruleUpdateCb(pods []*statesinformer.PodMeta) error {
...
 		corev1.PodQOSGuaranteed, corev1.PodQOSBurstable, corev1.PodQOSBestEffort} {
 		bvtValue := r.getKubeQOSDirBvtValue(kubeQOS)
-		kubeQOSCgroupPath := koordletutil.GetKubeQosRelativePath(kubeQOS)
-		if err := sysutil.CgroupFileWrite(kubeQOSCgroupPath, sysutil.CPUBVTWarpNs, strconv.FormatInt(bvtValue, 10)); err != nil {
+		kubeQOSCgroupPath := koordletutil.GetPodQoSRelativePath(kubeQOS)
+		bvtUpdater, err := resourceexecutor.NewCommonCgroupUpdater(sysutil.CPUBVTWarpNsName, kubeQOSCgroupPath, strconv.FormatInt(bvtValue, 10))
+		if err != nil {
+			klog.Infof("bvtupdater creat failed, dir %v, error %v", kubeQOSCgroupPath, err)
+		}
+		if err := bvtUpdater.Update(); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/groupidentity/rule.go:140]: set Update as private, use executor here

File: pkg/koordlet/runtimehooks/protocol/protocol.go [hunk 17]
```diff
@@ -101,7 +104,8 @@ func injectMemoryLimit(cgroupParent string, memoryLimit int64) error {
 	if err != nil {
 		return err
 	}
-	return updater.Update()
+	_, err = resourceexecutor.NewResourceUpdateExecutor().Update(false, updater)
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/protocol/protocol.go:107]: pass resourceexecutor into server

Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/protocol/protocol.go:107]: pass resourceexecutor into server

File: pkg/koordlet/runtimehooks/reconciler/reconciler.go [hunk 18]
```diff
@@ -204,12 +205,18 @@ type Reconciler interface {
 	Run(stopCh <-chan struct{}) error
 }
 
-func NewReconciler(s statesinformer.StatesInformer) Reconciler {
+type Options struct {
+	S        statesinformer.StatesInformer
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/reconciler/reconciler.go:209]: StatesInformer

File: pkg/koordlet/runtimehooks/protocol/protocol.go [hunk 19]
```diff
@@ -69,51 +69,51 @@ func (r *Resources) IsOriginResSet() bool {
...
 
-func injectCPUShares(cgroupParent string, cpuShares int64) error {
+func injectCPUShares(cgroupParent string, cpuShares int64, e resourceexecutor.ResourceUpdateExecutor) error {
 	cpuShareStr := strconv.FormatInt(cpuShares, 10)
 	updater, err := resourceexecutor.DefaultCgroupUpdaterFactory.New(sysutil.CPUSharesName, cgroupParent, cpuShareStr)
 	if err != nil {
 		return err
 	}
-	_, err = resourceexecutor.NewResourceUpdateExecutor().Update(false, updater)
+	_, err = e.Update(false, updater)
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/protocol/protocol.go:78]: true
